# **IRONHACK_Sharks**

In [18]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')  
pd.set_option('display.max_columns', None)  
from src.shark_func import *

Proyecto de limpieza del archivo attacks.csv, que se puede localizar en el siguiente link https://www.kaggle.com/datasets/teajay/global-shark-attacks/download?datasetVersionNumber=7

In [19]:
# attacks.csv no está codificado en UTF-8, hay que especificar la codificación en la importación
sharks = pd.read_csv('../data/attacks.csv',encoding='ANSI')    

In [42]:
shape_origin = sharks.shape         # Guardamos el tamaño inicial del dataset
sharks.head(7)

Case Number         Date    Year        Type    Country             Area   
0    2018.06.25  25-Jun-2018  2018.0     Boating        USA       California  \
1    2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2    2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3    2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4    2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
5  2018.06.03.b  03-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
6  2018.06.03.a  03-Jun-2018  2018.0  Unprovoked     BRAZIL       Pernambuco   

                         Location      Activity                    Name Sex    
0     Oceanside, San Diego County      Paddling             Julie Wolfe    F  \
1  St. Simon Island, Glynn County      Standing         Adyson McNeely     F   
2                    Habush, Oahu       Surfing             John Denges    M   
3              Arrawarra Headland       Surfing                    male    M   
4                        La Ticla   Free diving          Gustavo Ramos     M   
5              Flat Rock, Ballina  Kite surfing                 Chris …    M   
6           Piedade Beach, Recife      Swimming  Jose Ernesto da Silva     M   

   Age                                             Injury Fatal (Y/N)   
0   57  No injury to occupant, outrigger canoe and pad...           N  \
1   11                         Minor injury to left thigh           N   
2   48       Injury to left lower leg from surfboard skeg           N   
3  NaN                          Minor injury to lower leg           N   
4  NaN  Lacerations to leg & hand shark PROVOKED INCIDENT           N   
5  NaN                            No injury, board bitten           N   
6   18                                              FATAL           Y   

             Time         Species           Investigator or Source   
0           18h00      White shark                R. Collier, GSAF  \
1   14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2           07h45              NaN  K.McMurray, TrackingSharks.com   
3             NaN        2 m shark                  B. Myatt, GSAF   
4             NaN  Tiger shark, 3m                       A .Kipper   
5             NaN              NaN       Daily Telegraph, 6/4/2018   
6  Late afternoon      Tiger shark  Diario de Pernambuco, 6/4/2018   

                         pdf   
0       2018.06.25-Wolfe.pdf  \
1     2018.06.18-McNeely.pdf   
2      2018.06.09-Denges.pdf   
3   2018.06.08-Arrawarra.pdf   
4       2018.06.04-Ramos.pdf   
5  2018.06.03.b-FlatRock.pdf   
6   2018.06.03.a-daSilva.pdf   

                                        href formula   
0  http://sharkattackfile.net/spreadsheets/pdf_di...  \
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   
5  http://sharkattackfile.net/spreadsheets/pdf_di...   
6  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1   
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25  \
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   
5  http://sharkattackfile.net/spreadsheets/pdf_di...  2018.06.03.b   
6  http://sharkattackfile.net/spreadsheets/pdf_di...  2018.06.03.a   

  Case Number.2  original order Unnamed: 22 Unnamed: 23  
0    2018.06.25          6303.0     Unknown     Unknown  
1    2018.06.18          6302.0     Unknown     Unknown  
2    2018.06.09          6301.0     Unknown     Unknown  
3    2018.06.08          6300.0     Unknown     Unknown  
4    2018.06.04    

In [21]:
sharks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

#### PRIMERAS IMPRESIONES:
* **Case Number:** identificación única del caso (tip formatearlo de alguna manera para poder usar la fecha)
* **Date:** Fecha
* Year
* Type  Tiene pinta que es el ataque (value counts provoke unprovoke unknown)
* Country
* Area
* Location
* Name
* Sex
* Age
* Injury
* Fatal (yes or no/ Unknown)
* Time
* Species (aquí hay chica)
* Investigator or Source: posibilidad de pasarla a dos columnas
* pdf
* href formula
* href duplicada
* Case number 1
* Case number 2
* original order: WTF
* Unnamed 22: Las voy a usar para otras cosas
* Unnamed 23: Las voy a usar para otras cosas

Vamos a comprobar si tenemos filas duplicadas

In [22]:
sharks.duplicated().sum()

19411

Tenemos 19411 lineas duplicadas. Procedemos a eliminarlas. No aportan información alguna

In [24]:
sharks.drop_duplicates(inplace = True)
sharks.reset_index(drop=True,inplace=True)

In [26]:
sharks.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6312 entries, 0 to 6311
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             6310 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null   

In [ ]:
len(check_nan_cols(sharks))

In [ ]:
len(check_nan_cols(sharks,method='avg'))

### **1 - Eliminación de nulos**

Nuestro primer objetivo es que no haya ninguna columna con valores nulos. Se observa que todas las columnas tienen nulos. En algunas (Unnamed) prácticamente el 100%. Iremos analizando cada una, e iremos tomando decisiones para ir resolviendo los conflictos

#### **Columnas Unnamed**

In [27]:
# Hemos visto que solo hay 2 valores no nulos en Unnamed:23 y un valor no nulo en Unnamed:22

sharks[(sharks['Unnamed: 22'].notna()) | sharks['Unnamed: 23'].notna()]   # devuleve las filas en las que tengamos nulos en las columnas Unnamed 

Case Number                  Date    Year        Type   
1478    2006.05.27           27-May-2006  2006.0  Unprovoked  \
4415    1952.03.30           30-Mar-1952  1952.0  Unprovoked   
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0    Provoked   

                   Country         Area                    Location   
1478                   USA       Hawaii          North Shore, O'ahu  \
4415  NETHERLANDS ANTILLES      Curacao                         NaN   
5840                   USA  Connecticut  Branford, New Haven County   

                                             Activity              Name Sex    
1478                                          Surfing      Bret Desmond    M  \
4415  Went to aid of child being menaced by the shark       A.J. Eggink    M   
5840                                          Fishing  Captain Pattison    M   

      Age                                        Injury Fatal (Y/N)   Time   
1478   31             No injury, shark bumped surfboard           N  16h00  \
4415  NaN                Buttock bitten, tissue removed           N    NaN   
5840  NaN  Leg bitten by netted shark PROVOKED INCIDENT           N    NaN   

                                               Species    
1478                                                NaN  \
4415  Bull shark, 2.7 m [9'] was captured & dragged ...   
5840                                                NaN   

                                 Investigator or Source   
1478                                         R. Collier  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   
5840                       St. Joseph Herald, 9/14/1878   

                            pdf   
1478     2006.05.27-Desmond.pdf  \
4415      1952.03.30-Eggink.pdf   
5840  1878.09.14.R-Pattison.pdf   

                                           href formula   
1478  http://sharkattackfile.net/spreadsheets/pdf_di...  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1   
1478  http://sharkattackfile.net/spreadsheets/pdf_di...    2006.05.27  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...  1878.09.14.R   

     Case Number.2  original order   Unnamed: 22      Unnamed: 23  
1478    2006.05.27          4825.0  stopped here              NaN  
4415    1952.03.30          1888.0           NaN           Teramo  
5840  1878.09.14.R           463.0           NaN  change filename

Los valores no nulos que tenemos son: 'stopped here', 'Terrano' y 'change filename'. Información sin realción alguna con los ataques a tiburones. Vamos a optar añadir a estas dos columnas por 'Unknown'. Más adelante veremos si nos es conveniente usar estas columnas para albergar otra información que consideremos relevante


In [28]:
sharks[['Unnamed: 22', 'Unnamed: 23']] = ['Unknown','Unknown']

In [29]:
# Vuelvo a checkear duplicados en las filas

sharks.duplicated().sum()

0

In [30]:
check_nan_cols(sharks,);

Case Number                  2
Date                        10
Year                        12
Type                        14
Country                     60
Area                       465
Location                   550
Activity                   554
Name                       220
Sex                        575
Age                       2841
Injury                      38
Fatal (Y/N)                549
Time                      3364
Species                   2848
Investigator or Source      27
pdf                         10
href formula                11
href                        10
Case Number.1               10
Case Number.2               10
original order               3
dtype: int64

#### **Columna original order**

In [52]:
sharks['original order'].head(10)

0    6303.0
1    6302.0
2    6301.0
3    6300.0
4    6299.0
5    6298.0
6    6297.0
7    6296.0
8    6295.0
9    6294.0
Name: original order, dtype: float64

In [53]:
sharks['original order'].tail(10)

6302    6304.0
6303    6305.0
6304    6306.0
6305    6307.0
6306    6308.0
6307    6309.0
6308    6310.0
6309       NaN
6310       NaN
6311       NaN
Name: original order, dtype: float64

In [31]:
sharks['original order'].value_counts()  # Columna de enteros en la que todos son valores únicos menos 569 que se repite 1 vez

original order
569.0     2
6303.0    1
2102.0    1
2093.0    1
2094.0    1
         ..
4203.0    1
4204.0    1
4205.0    1
4206.0    1
6310.0    1
Name: count, Length: 6308, dtype: int64

In [50]:
sharks[sharks['original order']==569]

Case Number         Date    Year        Type       Country   
5733  1888.01.22  22-Jan-1888  1888.0     Boating     AUSTRALIA  \
5734  1888.02.00     Feb-1888  1888.0  Unprovoked  SOUTH AFRICA   

                       Area              Location                  Activity   
5733        New South Wales         Sydney Harbor                    Rowing  \
5734  Eastern Cape Province  Mzimvubu River mouth  Crossing the river mouth   

       Name Sex   Age                                             Injury   
5733  Burke    M  NaN  Shark bit boat, but no injury to occupant who ...  \
5734   male    M  NaN                                              FATAL   

     Fatal (Y/N) Time Species    
5733           N  NaN      NaN  \
5734           Y  NaN      NaN   

                                 Investigator or Source   
5733                                    Star, 1/23/1888  \
5734  Cape Mercantile Advertiser, 2/15/1888, M. Levi...   

                          pdf   
5733     1888.01.22-Burke.pdf  \
5734  1888.02.00-Mzimvubu.pdf   

                                           href formula   
5733  http://sharkattackfile.net/spreadsheets/pdf_di...  \
5734  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1   
5733  http://sharkattackfile.net/spreadsheets/pdf_di...    1888.01.22  \
5734  http://sharkattackfile.net/spreadsheets/pdf_di...    1888.02.00   

     Case Number.2  original order Unnamed: 22 Unnamed: 23  
5733    1888.01.22           569.0     Unknown     Unknown  
5734    1888.02.00           569.0     Unknown     Unknown

In [51]:
sharks['original order'].max() - sharks['original order'].min() 

6308.0

Esta columna representa el numero de ordenación original del dataset. Hemos llegado a la siguiente concluisión con estos datos:
* Vemos que registros consecutivos tienen numeros consecutivos en esta columna
* La diferencia entre el maximo y el minomo es de 6308, si le sumamos el valor repetido y los tres valores nulos suman 6312. La longitud de nuestro dataset
* SE LLAMA ORIGINAL ORDER!!!!!!

Como esta columna es unicamente para referenciar registros, vamos a ordenarla de 1 a 6312. Pero necesitamos saber como voy a tratar sus 3 nulos.

In [54]:
sharks[sharks['original order'].isna()]

Case Number Date  Year Type Country Area Location Activity Name Sex    
6309           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN  \
6310         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
6311          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

      Age Injury Fatal (Y/N) Time Species  Investigator or Source  pdf   
6309  NaN    NaN         NaN  NaN      NaN                    NaN  NaN  \
6310  NaN    NaN         NaN  NaN      NaN                    NaN  NaN   
6311  NaN    NaN         NaN  NaN      NaN                    NaN  NaN   

     href formula href Case Number.1 Case Number.2  original order   
6309          NaN  NaN           NaN           NaN             NaN  \
6310          NaN  NaN           NaN           NaN             NaN   
6311          NaN  NaN           NaN           NaN             NaN   

     Unnamed: 22 Unnamed: 23  
6309     Unknown     Unknown  
6310     Unknown     Unknown  
6311     Unknown     Unknown

Claramente estos 3 registros no aportan valor. Procedemos a eliminarlos

In [55]:
bad_index = sharks[sharks['original order'].isna()].index
sharks.drop(index=bad_index,inplace=True)

Comlpetamos esta columna en el rango de 1 a n. Esto nos puede dar información más adelante para completar fechas o registros

In [66]:
sharks['original order'].is_monotonic_increasing

False

In [58]:
sharks.loc[600:630,['original order']]

original order
600          5703.0
601          5702.0
602          5701.0
603          5700.0
604          5699.0
605          5698.0
606          5697.0
607          5696.0
608          5695.0
609          5694.0
610          5693.0
611          5692.0
612          5691.0
613          5690.0
614          5689.0
615          5688.0
616          5687.0
617          5686.0
618          5685.0
619          5684.0
620          5683.0
621          5682.0
622          5681.0
623          5680.0
624          5679.0
625          5678.0
626          5677.0
627          5676.0
628          5675.0
629          5674.0
630          5673.0

In [62]:
sharks.head(20)

Case Number                  Date    Year          Type       Country   
0     2018.06.25           25-Jun-2018  2018.0       Boating           USA  \
1     2018.06.18           18-Jun-2018  2018.0    Unprovoked           USA   
2     2018.06.09           09-Jun-2018  2018.0       Invalid           USA   
3     2018.06.08           08-Jun-2018  2018.0    Unprovoked     AUSTRALIA   
4     2018.06.04           04-Jun-2018  2018.0      Provoked        MEXICO   
5   2018.06.03.b           03-Jun-2018  2018.0    Unprovoked     AUSTRALIA   
6   2018.06.03.a           03-Jun-2018  2018.0    Unprovoked        BRAZIL   
7     2018.05.27           27-May-2018  2018.0    Unprovoked           USA   
8   2018.05.26.b           26-May-2018  2018.0    Unprovoked           USA   
9   2018.05.26.a           26-May-2018  2018.0    Unprovoked           USA   
10    2018.05.24           24-May-2018  2018.0      Provoked     AUSTRALIA   
11    2018.05.21           21-May-2018  2018.0    Unprovoked           USA   
12  2018.05.13.b           13-May-2018  2018.0    Unprovoked           USA   
13  2018.05.13.a           13-May-2018  2018.0       Invalid       ENGLAND   
14    2018.05.00              May 2018  2018.0      Provoked     AUSTRALIA   
15    2018.05.12           12-May-2018  2018.0    Unprovoked  SOUTH AFRICA   
16    2018.05.09           09-May-2018  2018.0  Questionable     AUSTRALIA   
17  2018.04.30.R  Reported 30-Apr-2018  2018.0    Unprovoked      THAILAND   
18  2018.04.28.b           28-Apr-2018  2018.0    Unprovoked    COSTA RICA   
19  2018.04.28.a           28-Apr-2018  2018.0    Unprovoked     AUSTRALIA   

                     Area                                           Location   
0              California                        Oceanside, San Diego County  \
1                 Georgia                     St. Simon Island, Glynn County   
2                  Hawaii                                       Habush, Oahu   
3         New South Wales                                 Arrawarra Headland   
4                  Colima                                           La Ticla   
5         New South Wales                                 Flat Rock, Ballina   
6              Pernambuco                              Piedade Beach, Recife   
7                 Florida  Lighhouse Point Park, Ponce Inlet, Volusia County   
8                 Florida                       Cocoa Beach, Brevard  County   
9                 Florida                      Daytona Beach, Volusia County   
10             Queensland                                    Cairns Aquarium   
11         South Carolina                  Isle of Palms,  Charleston County   
12         South Carolina                Hilton Head Island, Beaufort County   
13               Cornwall                                     Off Land's End   
14      Westerm Australia                                         Dugong Bay   
15  Eastern Cape Province                          Nahoon Beach, East London   
16        New South Wales                             Sharpes Beach, Ballina   
17                Hua Hin                                      Sai Noi Beach   
18           Cocos Island                                          Manuelita   
19      Western Australia                                       Gearys Beach   

           Activity                    Name Sex   Age   
0          Paddling             Julie Wolfe    F   57  \
1          Standing         Adyson McNeely     F   11   
2           Surfing             John Denges    M   48   
3           Surfing                    male    M  NaN   
4       Free diving          Gustavo Ramos     M  NaN   
5      Kite surfing                 Chris …    M  NaN   
6          Swimming  Jose Ernesto da Silva     M   18   
7           Fishing                    male    M   52   
8           Walking               Cody High    M   15   
9          Standing                    male    M   12   
10   Feeding sharks                    male    M   32   
1